File: frontend/src/modules/order/models/order.ts
/**
 * Order domain models and interfaces
 * Core order management types for the application
 */

import type { Address } from '@/modules/core/models/base/address';
import type { TenantScoped, UUID } from '@/modules/core/models/base';
import type { Money } from '@/modules/core/models/base/money';

/**
 * Order status enum representing the order lifecycle
 */
export enum OrderStatus {
  PENDING = 'PENDING',
  PAID = 'PAID',
  PROCESSING = 'PROCESSING',
  SHIPPED = 'SHIPPED',
  DELIVERED = 'DELIVERED',
  CANCELLED = 'CANCELLED',
  REFUNDED = 'REFUNDED',
  FAILED = 'FAILED',
}

/**
 * Order source enum representing different channels
 */
export enum OrderSource {
  WEBSITE = 'WEBSITE',
  MOBILE_APP = 'MOBILE_APP',
  WHATSAPP = 'WHATSAPP',
  INSTAGRAM = 'INSTAGRAM',
  FACEBOOK = 'FACEBOOK',
  PHONE = 'PHONE',
  IN_PERSON = 'IN_PERSON'
}

/**
 * Payment method enum for different payment options
 */
export enum PaymentMethod {
  CREDIT_CARD = 'CREDIT_CARD',
  DEBIT_CARD = 'DEBIT_CARD',
  CARD = 'CARD',
  PAYPAL = 'PAYPAL',
  BANK_TRANSFER = 'BANK_TRANSFER',
  CASH_ON_DELIVERY = 'CASH_ON_DELIVERY',
  MOBILE_MONEY = 'MOBILE_MONEY'
}

/**
 * Payment status enum for tracking payment state
 */
export enum PaymentStatus {
  PENDING = 'PENDING',
  PROCESSING = 'PROCESSING',
  PAID = 'PAID',
  COMPLETED = 'COMPLETED',
  FAILED = 'FAILED',
  REFUNDED = 'REFUNDED',
  CANCELLED = 'CANCELLED'
}

/**
 * Shipping method enum for different shipping options
 */
export enum ShippingMethod {
  STANDARD = 'STANDARD',
  EXPRESS = 'EXPRESS',
  OVERNIGHT = 'OVERNIGHT',
  PICKUP = 'PICKUP',
  RIDER = 'RIDER',
  COURIER = 'COURIER'
}

/**
 * Order item representing a product in an order
 */
export interface OrderItem {
  id: string;
  product_id: string;
  variant_id?: string;
  name: string;
  description?: string;
  quantity: number;
  unit_price: Money;
  total_price: Money;
  image_url?: string;
  sku?: string;
  weight?: number;
  weight_unit?: string;
  // Additional properties for return functionality
  product_name?: string;
  product_image?: string;
  variant_name?: string;
  returned_quantity?: number;
  created_at: string;
  updated_at: string;
}

/**
 * Shipping plugin meta information
 */
export interface ShippingPluginMeta {
  provider: string; // e.g. 'Sendy', 'DHL', 'CustomPluginName'
  pluginData?: Record<string, unknown>;
}

/**
 * Shipping details for an order
 */
export interface ShippingDetails {
  address: Address;
  method: ShippingMethod | string;
  pluginMeta?: ShippingPluginMeta;
  tracking_number?: string;
  estimated_delivery?: string;
  shipping_cost: Money;
  notes?: string;
}

/**
 * Payment details for an order
 */
export interface PaymentDetails {
  method: PaymentMethod;
  status: PaymentStatus;
  transaction_id?: string;
  provider?: string;
  amount_paid: Money;
  payment_date?: string;
  receipt_url?: string;
  last_four?: string; // For card payments
  phone_number?: string; // For mobile money
}

/**
 * Customer information for an order
 */
export interface CustomerInfo {
  id: string;
  name: string;
  email: string;
  phone: string;
  is_guest: boolean;
  created_at?: string;
  updated_at?: string;
}

/**
 * Order timeline entry for tracking order history
 */
export interface OrderTimeline {
  id: string;
  order_id: string;
  status: OrderStatus;
  timestamp: string;
  notes?: string;
  created_by?: string;
  created_at: string;
}

/**
 * Complete Order domain model with all details
 * Implements TenantScoped for multi-tenant isolation
 */
export interface Order extends TenantScoped {
  id: UUID;
  tenant_id: UUID;         // Merchant ID for tenant isolation
  order_number: string;
  customer: CustomerInfo;
  items: OrderItem[];
  total_amount: Money;
  subtotal: Money;
  tax: Money;
  status: OrderStatus;
  source: OrderSource;
  shipping: ShippingDetails;
  payment: PaymentDetails;
  notes?: string;
  metadata?: Record<string, unknown>;
  timeline: OrderTimeline[];
  idempotency_key: string;
  created_at: string;
  updated_at?: string; // Used to prevent duplicate orders
}

/**
 * Domain methods for the Order model
 */
export const OrderDomainMethods = {
  /**
   * Check if an order can be cancelled
   */
  canBeCancelled(order: Order): boolean {
    return [OrderStatus.PENDING, OrderStatus.PAID].includes(order.status);
  },

  /**
   * Check if an order can be refunded
   */
  canBeRefunded(order: Order): boolean {
    return (
      [OrderStatus.PAID, OrderStatus.PROCESSING, OrderStatus.SHIPPED].includes(order.status) &&
      order.payment.status === PaymentStatus.COMPLETED
    );
  },

  /**
   * Check if an order has been completed
   */
  isComplete(order: Order): boolean {
    return order.status === OrderStatus.DELIVERED;
  },

  /**
   * Calculate the total items in an order
   */
  getTotalItems(order: Order): number {
    return order.items.reduce((total, item) => total + item.quantity, 0);
  },

  /**
   * Get the latest timeline event
   */
  getLatestTimelineEvent(order: Order): OrderTimeline | undefined {
    if (!order.timeline.length) return undefined;

    return [...order.timeline].sort((a, b) => {
      return new Date(b.timestamp).getTime() - new Date(a.timestamp).getTime();
    })[0];
  },
};

File: backend/app/domain/models/order.py
import enum
import uuid
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator
from app.core.exceptions import AppError

# ============================================================
# Domain enums
# ============================================================


class OrderStatus(str, enum.Enum):
    PENDING = "PENDING"
    PAID = "PAID"
    PROCESSING = "PROCESSING"
    SHIPPED = "SHIPPED"
    DELIVERED = "DELIVERED"
    CANCELLED = "CANCELLED"
    REFUNDED = "REFUNDED"
    FAILED = "FAILED"


class OrderSource(str, enum.Enum):
    WHATSAPP = "WHATSAPP"
    WEBSITE = "WEBSITE"
    INSTAGRAM = "INSTAGRAM"


class PaymentMethod(str, enum.Enum):
    CARD = "CARD"
    MOBILE_MONEY = "MOBILE_MONEY"
    CASH_ON_DELIVERY = "CASH_ON_DELIVERY"
    BANK_TRANSFER = "BANK_TRANSFER"
    USSD = "USSD"


class PaymentStatus(str, enum.Enum):
    PENDING = "PENDING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"
    REFUNDED = "REFUNDED"
    PARTIALLY_REFUNDED = "PARTIALLY_REFUNDED"


# ============================================================
# Domain value objects
# ============================================================


class Money(BaseModel):
    amount: float
    currency: str = "KES"  # Default to Kenyan Shilling


class Coordinates(BaseModel):
    latitude: float
    longitude: float


class Address(BaseModel):
    street: str
    city: str
    state: str
    postal_code: Optional[str] = None
    country: str = "Kenya"  # Default to Kenya
    apartment: Optional[str] = None
    landmark: Optional[str] = None  # Important for African markets
    coordinates: Optional[Coordinates] = None


class CustomerInfo(BaseModel):
    id: Optional[str] = None  # Optional for guest checkout
    name: str
    # Email can be optional for WhatsApp orders
    email: Optional[EmailStr] = None
    phone: str
    is_guest: bool = True

    @field_validator("phone")
    @classmethod
    def validate_phone(cls, v):
        import re

        if not re.match(r"^\+?[0-9]{10,15}$", v):
            raise OrderDomainError("Invalid phone number format")
        return v


class OrderItem(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    product_id: str
    product_name: str
    quantity: int
    unit_price: Money
    total_price: Money
    variant_id: Optional[str] = None
    variant_name: Optional[str] = None
    image_url: Optional[str] = None


class ShippingDetails(BaseModel):
    address: Address
    method: str
    tracking_number: Optional[str] = None
    estimated_delivery: Optional[datetime] = None
    shipping_cost: Money
    notes: Optional[str] = None


class PaymentDetails(BaseModel):
    method: PaymentMethod
    status: PaymentStatus = PaymentStatus.PENDING
    transaction_id: Optional[str] = None
    provider: Optional[str] = None
    amount_paid: Money
    payment_date: Optional[datetime] = None
    receipt_url: Optional[str] = None
    last_four: Optional[str] = None  # For card payments
    phone_number: Optional[str] = None  # For mobile money


class OrderTimeline(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    status: OrderStatus
    timestamp: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc))
    note: Optional[str] = None
    created_by: Optional[str] = None


# ============================================================
# Domain model
# ============================================================


class Order(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    tenant_id: str
    order_number: str
    customer: CustomerInfo
    items: List[OrderItem]
    total_amount: Money
    subtotal: Money
    tax: Money
    status: OrderStatus = OrderStatus.PENDING
    source: OrderSource
    shipping: ShippingDetails
    payment: PaymentDetails
    notes: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None
    timeline: List[OrderTimeline] = []
    idempotency_key: str
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc))

    # Domain methods
    def can_be_cancelled(self) -> bool:
        """Check if an order can be cancelled"""
        return self.status in [OrderStatus.PENDING, OrderStatus.PAID]

    def can_be_refunded(self) -> bool:
        """Check if an order can be refunded"""
        return (
            self.status
            in [OrderStatus.PAID, OrderStatus.PROCESSING, OrderStatus.SHIPPED]
            and self.payment.status == PaymentStatus.COMPLETED
        )

    def is_complete(self) -> bool:
        """Check if an order has been completed"""
        return self.status == OrderStatus.DELIVERED

    def get_total_items(self) -> int:
        """Calculate the total items in an order"""
        return sum(item.quantity for item in self.items)

    def get_latest_timeline_event(self) -> Optional[OrderTimeline]:
        """Get the latest timeline event"""
        if not self.timeline:
            return None

        return sorted(self.timeline, key=lambda x: x.timestamp, reverse=True)[0]

    def add_timeline_event(
        self,
        status: OrderStatus,
        note: Optional[str] = None,
        created_by: Optional[str] = None,
    ) -> None:
        """Add a new timeline event to the order"""
        timeline_event = OrderTimeline(
            status=status, note=note, created_by=created_by)
        self.timeline.append(timeline_event)
        self.status = status
        self.updated_at = datetime.now(timezone.utc)


# ============================================================
# Request/Response models
# ============================================================


class CreateOrderRequest(BaseModel):
    customer: CustomerInfo
    items: List[OrderItem]
    shipping: ShippingDetails
    payment: PaymentDetails
    source: OrderSource
    notes: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None
    idempotency_key: str = Field(default_factory=lambda: str(uuid.uuid4()))
    tenant_id: str

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "customer": {
                    "name": "John Doe",
                    "email": "john@example.com",
                    "phone": "+254712345678",
                    "is_guest": True,
                },
                "items": [
                    {
                        "product_id": "550e8400-e29b-41d4-a716-446655440000",
                        "product_name": "Smartphone",
                        "quantity": 1,
                        "unit_price": {"amount": 45000, "currency": "KES"},
                        "total_price": {"amount": 45000, "currency": "KES"},
                    }
                ],
                "shipping": {
                    "address": {
                        "street": "123 Main St",
                        "city": "Nairobi",
                        "state": "Nairobi",
                        "country": "Kenya",
                        "landmark": "Near Central Park",
                    },
                    "method": "Standard Delivery",
                    "shipping_cost": {"amount": 500, "currency": "KES"},
                },
                "payment": {
                    "method": "MOBILE_MONEY",
                    "amount_paid": {"amount": 45500, "currency": "KES"},
                    "phone_number": "+254712345678",
                },
                "source": "WEBSITE",
                "tenant_id": "550e8400-e29b-41d4-a716-446655440000",
            }
        }
    )


class OrderResponse(BaseModel):
    order: Order


class OrdersResponse(BaseModel):
    orders: List[Order]
    total: int
    page: Optional[int] = None
    limit: Optional[int] = None


class OrderDomainError(AppError):
    pass
